### ISSUE TITLE
Change semantics of unsigned division by zero

### ISSUE THREAD

**Issue by TrevorHansen** [stp/stp#223]
Unless anyone has objections, I'm going to change the semantics of unsigned division by zero to return  -1 (all ones). Currently, we return 1 for unsigned division by zero if the input language is SMTLIB or SMTLIB2, but not by default for the APIs or CVC. 

Here's a related discussion.
http://www.cs.nyu.edu/pipermail/smt-lib/2015/000966.html

**Comment by TrevorHansen** [comment]
Some background. In the distant past, STP only read CVC and had an API. I believe that neither defined the semantics of division-by-zero. SMTLIB1 defined the semantics as undefined, which was too painful to implement, so for SMTLIB1 & SMTLIB2 I arbitrarily defined division-by-zero to equal 1. Choosing this for division-by-zero, also defines what what signed-division-by-zero does. 

To preserve backward compatibility, I added the API function "make_division_total" which gave the semantics of x/0 = 1, via the API. CVC was never changed. Effectively if you use the API or CVC format, it's your job to not put division-by-zeroes in the input.

SMTLIB seems to be soon changing the semantics for division-by-zero to return -1. It's not final yet though, some people things x/0 should equal 0.

As a step towards the new semantics, I've submitted 01ec223929eff58358499c739da0efe56dd3ff7c which turns on the "x/0" =1 semantics for everyone.

**Comment by msoos** [comment]
Wow thanks for that work! The biggest pain I had was that the constant bit
propagation was exiting out on div by 0 and even if the constant value was
not used for anything meaningful, it still propagated it hence crashing.
Thanks for fixing that up!

**Comment by msoos** [comment]
Yeah, please do turn it into -1 and then we can close this issue for good! :)

**Comment by TrevorHansen** [comment]
Hiya Mate, my reading is that the change to the hasn't been confirmed yet. Because signed division is defined in terms of unsigned division, we need to change signed-division-by-zero too. I'm not sure what the signed-division-by-zero semantics will be, so was thinking of waiting until it was confirmed..

**Comment by msoos** [comment]
Well. Currently, it's undefined in the SMT definition. So If we define it to be -1, it's according to def (i.e. it's undefined so we can make it anything, in this case, -1). I think it's worse not having it defined than having it defined -1. With -1 I can write tests, STP can be tested to be at least deterministic and non-assert-failing/segfaulting/etc. Otherwise, we'll have this thing that makes the solver undefined. It's really not that good. How about fixing it to -1 now and then we can fix it to something else later? I guess changing it from -1 won't be a big deal. We can even just define it as a constant/commend-line-argument. Actually, latter could be interesting, as it would allow users to run the query 2x with two different definitions of div-by-zero. If the results are different, then div-by-zero mattered? Could be useful perhaps?

**Comment by TrevorHansen** [comment]
Hiya Mate, as of a couple of weeks ago, division by zero evaluates to 1. So it should be deterministic (and never crash) now. 

What they mean in the specification by undefined is that it can evaluate to anything, but functional congruence (i.e. same arguments evaluate to same result) applies. So with current SMTLLIB undefined semantics:  (2/0 = 5) must evaluate to satisfiable. So it's not correct, even though it's undefined, for us to evaluate 2/0 to 1. We do it anyway though, because doing undefined is a pain. Especially because signed division is defined in terms of unsigned division, so relationships between signed-division-by-zero and unsigned-division-by-zero need to be enforced.

Seems easiest to me to wait until the semantics are redefined, then implement/test carefully.

**Comment by msoos** [comment]
"We do it anyway though, because doing undefined is a pain" -> yes to that. I won't do any undefined crap, I'm on with you 100%. Yeah, let's wait until semantics are defined. Until then, I'm all good with the solution we have here.

I think we can close this? I'm closing and we can open another one when things get defined. I don't want this issue on the list: we are blocked by others to fix things.

**Comment by TrevorHansen** [comment]
The semantics for division & remainder by zero are about to be changed. I've started to implement the changed semantics. Relevant discussion is here:

http://www.cs.nyu.edu/pipermail/smt-lib/2017/001148.html

**Comment by TrevorHansen** [comment]
I've checked in a first version ( 34d453f71947c208315d555a4c9b64d2776ee61a ). I'll leave this case open until I've tested it properly.

**Comment by TrevorHansen** [comment]
I've tested and now believe that the semantics of bvudiv-by-zero evaluating to -1 are correctly implemented.
